### PR TITLE
Bug 1109684 - Let other developers directly open pages in Firefox for iOS

### PR DESCRIPTION
**Pull request by bkmunar**
1. Edited AppDelegate in order to handle custom URL schemes and created a reference to TabManager so that these links could be opened

### CODE REVIEW

File: Client/Application/AppDelegate.swift [hunk 1]
```diff
@@ -81,6 +82,23 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         return true
     }
 
+    func application(application: UIApplication, openURL url: NSURL, sourceApplication: String?, annotation: AnyObject?) -> Bool {
+        if url.scheme == "firefox" {
+            if url.lastPathComponent == "open-url" {
+                let query = url.query
```
Comment (fluffyemily) [inline Client/Application/AppDelegate.swift]: You're better off using if let here rather than checking for != nil
`if let query = url.query {
  ...
}`

File: Client/Application/AppDelegate.swift [hunk 2]
```diff
@@ -81,6 +82,23 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+                if query != nil {
+                    let index = query!.unescape().rangeOfString("=")
```
Comment (fluffyemily) [inline Client/Application/AppDelegate.swift]: same here. Nest this inside an `if let index = query!.unescape().rangeOfString("=") {...}`

**Comment by fluffyemily** [comment]
Just a couple of small swiftifications but otherwise good stuff :)

File: Client.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -1258,6 +1257,7 @@
 		59A68CCB63E2A565CB03F832 /* SearchViewController.swift */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.swift; path = SearchViewController.swift; sourceTree = "<group>"; };
 		6BB2FD971B017DAB001A189B /* AuralProgressBar.swift */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.swift; path = AuralProgressBar.swift; sourceTree = "<group>"; };
 		74203E241B276B3D007D481D /* SessionRestoreHandler.swift */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.swift; path = SessionRestoreHandler.swift; sourceTree = "<group>"; };
+		74657A291B2F7B0800B2FA9C /* OpenInFirefoxHelper.swift */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.swift; path = OpenInFirefoxHelper.swift; sourceTree = "<group>"; };
```
Comment (st3fan) [inline Client.xcodeproj/project.pbxproj]: What is this `OpenInFirefoxHelper.swift` file? I don't actually see it in the project.

File: Client.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -2924,6 +2925,7 @@
 				TargetAttributes = {
 					2827315D1ABC9BE600AA1954 = {
 						CreatedOnToolsVersion = 6.2;
+						DevelopmentTeam = 43AQ936H96;
```
Comment (st3fan) [inline Client.xcodeproj/project.pbxproj]: Don't commit this.

File: Client.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -5596,8 +5600,8 @@
 				CLANG_WARN_OBJC_ROOT_CLASS = YES_ERROR;
 				CLANG_WARN_UNREACHABLE_CODE = YES;
 				CLANG_WARN__DUPLICATE_METHOD_MATCH = YES;
-				CODE_SIGN_IDENTITY = "iPhone Distribution";
-				"CODE_SIGN_IDENTITY[sdk=iphoneos*]" = "iPhone Distribution";
+				CODE_SIGN_IDENTITY = "iPhone Developer";
+				"CODE_SIGN_IDENTITY[sdk=iphoneos*]" = "iPhone Developer";
```
Comment (st3fan) [inline Client.xcodeproj/project.pbxproj]: Don't commit any of these code signing changes please. If no new files were added to the project then you can probably remove this file from the commit completely.

File: Client/Application/AppDelegate.swift [hunk 6]
```diff
@@ -33,7 +34,7 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 
         let defaultRequest = NSURLRequest(URL: AppConstants.AboutHomeURL)
         let tabManager = TabManager(defaultNewTabRequest: defaultRequest, prefs: profile.prefs)
-
+        self.tabManager = tabManager
```
Comment (st3fan) [inline Client/Application/AppDelegate.swift]: You can drop the local instance on line 36 and just immediately do `tabManager = TabManager(...)`

File: Client/Application/AppDelegate.swift [hunk 7]
```diff
@@ -81,6 +82,22 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         return true
     }
 
+    func application(application: UIApplication, openURL url: NSURL, sourceApplication: String?, annotation: AnyObject?) -> Bool {
+        if url.scheme == "firefox" {
+            if url.lastPathComponent == "open-url" {
+                if let query = url.query {
+                    if let index = query.unescape().rangeOfString("=") {
```
Comment (st3fan) [inline Client/Application/AppDelegate.swift]: Instead of doing this with `rangeOfString`, use the `NSURLComponents` class. That will take a string and will parse the URL and and query for you.

File: Client/Application/AppDelegate.swift [hunk 8]
```diff
@@ -81,6 +82,22 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+                        let newRequest = NSURLRequest(URL: NSURL(string: query.substringFromIndex(index.endIndex))!)
```
Comment (st3fan) [inline Client/Application/AppDelegate.swift]: The forced unwrapping of `NSURL(string: ...)` is problematic. If another app sends a bad URL, then `NSURL(string:)` will return `nil` and that will cause a crash when it is passed to `NSURLRequest(URL:)`. Every time you force unwrap (using `!`) you have to ask yourself, can this operation fail or am I 100% sure that it is safe. If you are not sure then use a conditional like `if let` to make the code more robust.

File: Client/Info.plist [hunk 9]
```diff
@@ -64,7 +75,5 @@
 		<string>UIInterfaceOrientationLandscapeLeft</string>
 		<string>UIInterfaceOrientationLandscapeRight</string>
 	</array>
-	<key>UIViewControllerBasedStatusBarAppearance</key>
-	<true/>
```
Comment (st3fan) [inline Client/Info.plist]: Why was this removed?

Comment (bkmunar) [inline Client/Info.plist]: I think it just got moved to another row? (i don't think I did this), but if you open up the hidden lines there's a duplicate of it right under the UISUpportednterfaceOrientations stuff

Comment (st3fan) [inline Client/Info.plist]: Ahh I see. Looks like Xcode just cleaned it up. No worries then :-)

File: Client/Application/AppDelegate.swift [hunk 10]
```diff
@@ -78,6 +78,27 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        return false
+    }
+
```
Comment (bkmunar) [inline Client/Application/AppDelegate.swift]: yeahhh this got kind of convoluted when i tried to do the way you proposed!

File: Client.xcodeproj/project.pbxproj [hunk 11]
```diff
@@ -280,7 +280,6 @@
 		59A68E0B4ABBF55E14819668 /* BookmarksPanel.swift in Sources */ = {isa = PBXBuildFile; fileRef = 59A6839879D615FC1C0D71CE /* BookmarksPanel.swift */; };
 		59A68FD5260B8D520F890F4A /* ReaderPanel.swift in Sources */ = {isa = PBXBuildFile; fileRef = 59A685F4EAD19EDEC854BCA4 /* ReaderPanel.swift */; };
 		6BB2FD981B017DAB001A189B /* AuralProgressBar.swift in Sources */ = {isa = PBXBuildFile; fileRef = 6BB2FD971B017DAB001A189B /* AuralProgressBar.swift */; };
-		6BB2FD991B017DAB001A189B /* AuralProgressBar.swift in Sources */ = {isa = PBXBuildFile; fileRef = 6BB2FD971B017DAB001A189B /* AuralProgressBar.swift */; };
```
Comment (st3fan) [inline Client.xcodeproj/project.pbxproj]: Why is this removed? This has nothing to do with your patch?

File: Client.xcodeproj/project.pbxproj [hunk 12]
```diff
@@ -4282,8 +4280,8 @@
 				CLANG_WARN_OBJC_ROOT_CLASS = YES_ERROR;
 				CLANG_WARN_UNREACHABLE_CODE = YES;
 				CLANG_WARN__DUPLICATE_METHOD_MATCH = YES;
-				CODE_SIGN_IDENTITY = "iPhone Developer";
-				"CODE_SIGN_IDENTITY[sdk=iphoneos*]" = "iPhone Developer";
+				CODE_SIGN_IDENTITY = "iPhone Distribution";
+				"CODE_SIGN_IDENTITY[sdk=iphoneos*]" = "iPhone Distribution";
```
Comment (st3fan) [inline Client.xcodeproj/project.pbxproj]: Don't commit this :-)

File: Client/Application/AppDelegate.swift [hunk 13]
```diff
@@ -78,6 +78,27 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+                                if let newURL = NSURL(string: newURLComponents.value!) {
+                                    let tab = self.tabManager.addTab(request: NSURLRequest(URL: newURL))
+                                    self.tabManager.selectTab(tab)
+                                    return true
+                                }
+                            }
+                        }
+                    }
+                }
+            }
```
Comment (st3fan) [inline Client/Application/AppDelegate.swift]: Great. This is pretty fail-safe I think. I'm fine with it, but if you want to simplify this `if let pyramid` .. since Swift 1.2 you can do this:

```
if let components = NSURLComponents(URL: url, resolvingAgainstBaseURL: false) where components.scheme == "firefox" && components.host == "open-url" {
    if let query = components.query, item =  components.queryItems?.first as? NSURLQueryItem {
        ... 
    }
}
```
So you can combine multiple `if let` conditionals. The expression will fail as a whole as soon as one check fails so it is safe to use previously unwrapped values. The `where` clause lets you do some basic conditional changes at the end.

I'm fine with how this looks now because it is pretty clear, but if you want to bring the number of lines down then you can use the above technique :-)

Comment (bkmunar) [inline Client/Application/AppDelegate.swift]: I shall!! thanks for all your help !!!

File: Client.xcodeproj/project.pbxproj (hunk 11 above)
Comment (bkmunar) [inline Client.xcodeproj/project.pbxproj]: i think xcode did it? or something? i think it's a duplicate that it got rid of?

File: Client/Info.plist (hunk 9 above)
Comment (bkmunar) [inline Client/Info.plist]: lttm

**Comment by bkmunar** [comment]
lttm

**Comment by bkmunar** [comment]
lttm

**Comment by bkmunar** [comment]
lttm
